src/persist-txn/src/txns.rs

```diff
@@ -0,0 +1,408 @@
+/// `(ShardId, <empty>)` pair. The data may not be read before the timestamp of
+/// the update (which starts at the time it was initialized, but it may later be
+/// forwarded).
+///
+/// Example (building on both of the above):
+///
+/// ```text
+/// // d0 and d1 were both initialized before they were used above
+/// (d0, <empty>, 0, 1)
+/// (d1, <empty>, 2, 1)
+/// ```
+#[derive(Debug)]
+pub struct TxnsHandle {
+    pub(crate) txns_cache: TxnsCache,
+    pub(crate) txns_write: WriteHandle<ShardId, String, u64, i64>,
+    pub(crate) datas: DataHandles,
+}
+
+impl TxnsHandle {
+    /// Returns a [TxnsHandle] committing to the given txn shard.
+    ///
+    /// `txns_id` identifies which shard will be used as the txns WAL. MZ will
+    /// likely have one of these per env, used by all processes and the same
+    /// across restarts.
+    ///
+    /// This also does any (idempotent) initialization work: i.e. ensures that
+    /// the txn shard is readable at `init_ts` by appending an empty batch, if
+    /// necessary.
+    pub async fn open(init_ts: u64, client: PersistClient, txns_id: ShardId) -> Self {
+        let (mut txns_write, txns_read) = client
+            .open(
+                txns_id,
+                Arc::new(TodoSchema::<ShardId>::default()),
+                Arc::new(StringSchema),
+                Diagnostics {
+                    shard_name: "txns".to_owned(),
+                    handle_purpose: "commit txns".to_owned(),
+                },
+            )
+            .await
+            .expect("txns schema shouldn't change");
+        let txns_cache = TxnsCache::init(init_ts, txns_read, &mut txns_write).await;
+        TxnsHandle {
+            txns_cache,
+            txns_write,
+            datas: DataHandles {
+                client,
+                data_write: BTreeMap::new(),
+            },
+        }
+    }
+
+    /// Returns a new, empty transaction that can involve the data shards
+    /// registered with this handle.
+    pub fn begin(&self) -> Txn {
+        // TODO(txn): This is a method on the handle because we'll need
+        // WriteHandles once we start spilling to s3.
+        Txn::new()
+    }
+
+    /// Registers a data shard for use with this txn set.
+    ///
+    /// The data shard will be initially readable at `register_ts`. If
+    /// necessary, this method will append empty updates to make that happen. If
+    /// it is not possible to register the data at that time, an Err will be
+    /// returned with the minimum time the data shard could be registered. If
+    /// the data has already been registered, this method is idempotent.
+    ///
+    /// **WARNING!** While a data shard is registered to the txn set, writing to
+    /// it directly (i.e. using a WriteHandle instead of the TxnHandle,
+    /// registering it with another txn shard) will lead to incorrectness,
+    /// undefined behavior, and (potentially sticky) panics.
+    pub async fn register(
```

Comment:
If I want to execute N transactions that all involve shard S, then do I need to register S 1 time or N times?

Reply:
discussed this offline (once). ditto the above, i'm hoping that the added detail in the # Intuition and Jargon section in this latest revision makes this more obvious?